features/graph/api/src/main/java/org/opennms/netmgt/graph/api/generic/GenericEdge.java

```diff
@@ -32,34 +32,63 @@
 import java.util.Map;
 import java.util.Objects;
 
-import org.opennms.netmgt.graph.api.AbstractEdge;
+import org.opennms.netmgt.graph.api.Edge;
 import org.opennms.netmgt.graph.api.VertexRef;
 
-public class GenericEdge extends AbstractEdge<GenericVertex> {
+import com.google.common.base.MoreObjects;
 
-    private Map<String, Object> properties = new HashMap<>();
+public class GenericEdge extends GenericElement implements Edge {
+
+    private final GenericVertex source;
+    private final GenericVertex target;
 
     public GenericEdge(GenericVertex source, GenericVertex target) {
-        this((VertexRef) source, (VertexRef) target);
+        this(source, target, new HashMap<>());
+    }
+
+    public GenericEdge(GenericVertex source, GenericVertex target, Map<String, Object> properties) {
+        super(properties);
+        this.source = Objects.requireNonNull(source);
+        this.target = Objects.requireNonNull(target);
+        this.setNamespace(source.getNamespace());
+        this.setId(source.getId() + "->" + target.getId());
     }
 
-    public GenericEdge(VertexRef source, VertexRef target) {
-        super(source, target);
-        properties.put(GenericProperties.LABEL, getLabel());
-        properties.put(GenericProperties.NAMESPACE, getNamespace());
-        properties.put(GenericProperties.ID, getId());
+    /** Copy constructor */
+    public GenericEdge(GenericEdge copyMe) {
+        this(new GenericVertex(copyMe.source), new GenericVertex(copyMe.target), new HashMap<>(copyMe.getProperties()));
```

Comment:
Either access the properties via name or method, but don't mix.

Reply:
fixed